### ISSUE TITLE
[BUG] model.fit() overwrites user-defined state_names when fitting on integer data

### ISSUE THREAD

**Issue by shostykovich21** [pgmpy/pgmpy#2207]
**Describe the bug**

The model.fit() method destructively overwrites CPDs that contain user-defined state_names.
If a user defines a CPD with string state names (e.g., {'A': ['expert_knowledge', 'default']}) and then fits the model on data where 'A' is represented by integers (0, 1), the original state_names are permanently discarded and replaced by the integers.

While a warning about replacing the CPD is shown, it's not obvious that this will also destroy the state name mapping. This leads to two main problems:
--Loss of Information: The model loses the connection between the integers and their string labels.
--Failed Queries: Subsequent inference using the original string labels (e.g., evidence={'A': 'expert_knowledge'}) will fail.

The expected behavior is for .fit() to use the existing state_names to interpret the numerical data, rather than deleting them. This aligns with the common practice in data science libraries of preserving, rather than overwriting, user-defined model information during fitting.
This would prevent users from having to manually pre-process their data to avoid this issue.

**To Reproduce**

```python
import pandas as pd
from pgmpy.models import DiscreteBayesianNetwork
from pgmpy.factors.discrete import TabularCPD
from pgmpy.estimators import MaximumLikelihoodEstimator

# Define a model with an expert-defined CPD for 'A'
# The states for 'A' are meaningful strings.
model = DiscreteBayesianNetwork([('A', 'B')])
cpd_a = TabularCPD(
    variable='A', variable_card=2,
    values=[[0.7], [0.3]],
    state_names={'A': ['expert_knowledge', 'default']}
)
model.add_cpds(cpd_a)

# Capture the correct, original state names before fitting
original_a_state_names = model.get_cpds('A').state_names['A']

# Create data where 'A' is represented by integers
data = pd.DataFrame(data={
    'A': [1, 0, 1, 1, 0],
    'B': ['x', 'y', 'x', 'y', 'x']
})

# Fit the model. This is where the state names are overwritten.
print("--- Fitting model on integer data ---")
model.fit(data, estimator=MaximumLikelihoodEstimator)
print("--- Fit complete ---\n")

# Get the newly learned CPD for 'B' and show the state name inconsistency
cpd_b_learned = model.get_cpds('B')

print("Original state names for 'A' (before fit):")
print(original_a_state_names)
print("\nState names for 'A' from the new P(B|A)'s perspective (after fit):")
print(cpd_b_learned.state_names['A'])
```

**Expected behavior**
The 'fit()' method should be non-destructive. When it encounters a variable that already has a user-defined CPD with 'state_names', it should:
1.  **Not** replace the existing CPD or its 'state_names'.
2.  Use the existing 'state_names' list to interpret the integer data. For example, it should map the integer '0' in the data to the state 'expert_knowledge' and 1 to 'default'.
3. The newly learned CPD for the child node (P(B|A)) should correctly reference its parent's string-based states.

**Current behavior**

WARNING:pgmpy:Replacing existing CPD for A

Original state names for 'A' (before fit):
['expert_knowledge', 'default']

State names for 'A' from the new P(B|A)'s perspective (after fit):
[0, 1]

**Proposed fix**

The logic within model.fit() and the MaximumLikelihoodEstimator should be modified. If pre-defined state_names exist for a variable, the estimator should attempt to map the data to these names. If the mapping is successful, it should use the mapped states and skip the step of inferring states from the data. If the mapping seems unreasonable(for example, if the data contains values like 0 and 2 when the model was only defined with two states), it can revert to its current behavior.

I am currently working on a fix for this issue and plan to submit a pull request soon.

**Comment by ankurankan** [comment]
@shostykovich21 I don't think matching is a good approach here, since we can't reliably tell whether the user intends to define new states or refer to existing ones. I'd suggest instead treating any new state names in the data as genuinely new states for the variable and updating the CPD to include the new states as well.

**Comment by shostykovich21** [comment]
> [@shostykovich21](https://github.com/shostykovich21) I don't think matching is a good approach here, since we can't reliably tell whether the user intends to define new states or refer to existing ones. I'd suggest instead treating any new state names in the data as genuinely new states for the variable and updating the CPD to include the new states as well.

I see. Nevertheless, if there's any alternative you'd prefer or something you'd like me to adjust in the current behavior, do let me know... I’d be happy to help!
